### PR TITLE
Add covering index support for PostgreSQL 11+

### PR DESCRIPTION
**Pull request by sebastian-palma**
This PR introduces the support for [covering indexes](https://www.postgresql.org/docs/11/indexes-index-only-scans.html) when using PostgreSQL 11 or newer.

The method `add_index` has been extended to allow an `includes` option, which
receives a single column or an array of columns:

    add_index :posts, :title, includes: :subtitle

Which produces a database index as:

    CREATE INDEX index_posts_on_title_subtitle ON posts(title) INCLUDE (subtitle);

The method `remove_index` has no changes, since it isn't needed to specify the includes columns to delete the index from the database, as we can use just the name (like with other index types).

There's a guard clause that checks if the `options` when invoking `add_index` contains `includes`, if so it checks if the PostgreSQL version is less than `110000` which is the value and format returned for versions greater or equal to 11. If the condition is true it raises a `NotImplementedError` error to warn the user the feature isn't available for the current PostgreSQL version used.

The clause is `INCLUDE` but the code was developed making use of `includes` as `include` is an already defined statement.

### CODE REVIEW

File: activerecord/lib/active_record/connection_adapters/abstract/schema_statements.rb [hunk 1]
```diff
@@ -1176,13 +1188,20 @@ def update_table_definition(table_name, base) #:nodoc:
       end
 
       def add_index_options(table_name, column_name, comment: nil, **options) # :nodoc:
+        if options.key?(:includes)
+          if ActiveRecord::Base.connection.select_value("SELECT CURRENT_SETTING('server_version_num')::INT < 110000")
```
Comment (yahonda) [inline activerecord/lib/active_record/connection_adapters/abstract/schema_statements.rb]: You can get PostgreSQL version by using `database_version` like https://github.com/rails/rails/blob/01336b71af6ac798e48101b4ceb2de00397243aa/activerecord/lib/active_record/connection_adapters/postgresql_adapter.rb#L204-L206

Comment (vnhnhm) [reply activerecord/lib/active_record/connection_adapters/abstract/schema_statements.rb]: Didn't think there was a method for this. I just pushed the commit with the changes.

Thanks for the feedback!

**Comment by tjwallace** [comment]
Does `remove_index` need to support the `includes` option so it can be reversed?

**Comment by vnhnhm** [comment]
> Does `remove_index` need to support the `includes` option so it can be reversed?

As far as I know not @tjwallace. The index can be removed by its name.

Not sure if that answers your question.

**Comment by tjwallace** [comment]
@vnhnhm what happens if you rollback a migration that contains `remove_index`? Will the index be rebuilt with the same `includes` values that the index had before the migration?

**Comment by vnhnhm** [comment]
Now I got it @tjwallace. It's throwing an `ArgumentError` error. I'm gonna work on that.

Thanks!

**Comment by vnhnhm** [comment]
Now it's working. Added the needed `add_index` revert logic and test cases @tjwallace.

File: activerecord/lib/active_record/connection_adapters/sqlite3_adapter.rb [hunk 2]
```diff
@@ -116,6 +116,10 @@ def supports_savepoints?
         true
       end
 
+      def supports_covering_indexes?
```
Comment (simi) [inline activerecord/lib/active_record/connection_adapters/sqlite3_adapter.rb]: Is this still needed?

Comment (sebastian-palma) [reply activerecord/lib/active_record/connection_adapters/sqlite3_adapter.rb]: It seems no @simi. It never was. I just rebased and removed that method from that class.

Thanks!

**Comment by sebastian-palma** [comment]
Any chance to review this @kamipo?

**Comment by simi** [comment]
Let's revive this. @sebastian-palma any chance to rebase this? I can take over if needed.

**Comment by sebastian-palma** [comment]
Hey @simi. Sure, I'm gonna take a look at it this weekend, it seems there's a lot to do.

Thanks!

**Comment by sebastian-palma** [comment]
> Let's revive this. @sebastian-palma any chance to rebase this? I can take over if needed.

I just updated the rebased code @simi ;)

**Comment by owen2345** [comment]
Hey @sebastian-palma 
As it is taking too much time, could you pls make a copy of the PR and create a gem with it?
Regards!

**Comment by sebastian-palma** [comment]
Hey @owen2345, I was thinking of something similar. I'll see if there's an active gem where I can add these changes, the best I can think right now is [GeorgeKaraszi/ActiveRecordExtended](https://github.com/GeorgeKaraszi/ActiveRecordExtended), but it seems it's far from the scope of the gem.

Anyway, I'll make the try. If you happen to know another gem where we can add this I could start working there.

**Comment by sebastian-palma** [comment]
I'm closing this because as I see there are no intentions to include this feature. Thanks everybody.

**Comment by fatkodima** [comment]
@sebastian-palma this is a great addition.

Please, reopen. Sometimes, PRs can take several years to get merged. You can verify this by looking at some of my commits here.

**Comment by sebastian-palma** [comment]
Hey @fatkodima, glad you find it useful, but it seems to me these changes are out of the scope of ActiveRecord. At some point I read it's not intended to replicate every feature every RDBMS has. As pointed by @owen2345, this could be a gem of its own or be added to some other library that incorporates new features to ActiveRecord, but not in this repo.

Of course if you want, we can reopen it for you to work and see what happens. It's all yours ;)

**Comment by sailor** [comment]
Hey,
It seems that something similar has just landed in Django. I, for sure, have some use cases where it would have come in handy. Thank's for the work anyway.

**Comment by typhoon2099** [comment]
We're currently having performance issues with a Rails app that covering indexes solve, but we have no way of adding to Rails as the schema.rb file does not correctly define them, so existing instances can get migrated fine, but new instances won't be indexed properly. This PR would be a massive improvement for our project.

**Comment by lxfontes** [comment]
hello from 2021.
if you got here via google, what you are looking for is https://gitlab.com/schlock/activerecord-covering-index

**Pull request closed without merging**